storage/series/buffer.go

```diff
@@ -626,26 +625,33 @@ func (b *dbBufferBucket) write(
 	return nil
 }
 
-func (b *dbBufferBucket) streams(ctx context.Context) []xio.SegmentReader {
+func (b *dbBufferBucket) streams(ctx context.Context) []xio.BlockReader {
 	// NB(r): Ensure we don't call any closers before the operation
 	// started by the passed context completes.
 	b.ctx.DependsOn(ctx)
 
-	streams := make([]xio.SegmentReader, 0, len(b.bootstrapped)+len(b.encoders))
+	streams := make([]xio.BlockReader, 0, len(b.bootstrapped)+len(b.encoders))
 
 	for i := range b.bootstrapped {
 		if b.bootstrapped[i].Len() == 0 {
 			continue
 		}
-		if s, err := b.bootstrapped[i].Stream(ctx); err == nil && s != nil {
+		if s, err := b.bootstrapped[i].Stream(ctx); err == nil && s.IsNotEmpty() {
 			// NB(r): block stream method will register the stream closer already
 			streams = append(streams, s)
 		}
 	}
 	for i := range b.encoders {
+		start := b.start
+		end := start.Add(b.opts.RetentionOptions().BlockSize())
 		if s := b.encoders[i].encoder.Stream(); s != nil {
-			ctx.RegisterFinalizer(s)
-			streams = append(streams, s)
+			br := xio.BlockReader{
+				SegmentReader: s,
+				Start:         start,
+				End:           end,
+			}
+			ctx.RegisterFinalizer(br)
```

Comment:
Can we register finalizer on the segment reader here too to avoid alloc on the interface?